distribution/lib/Standard/AWS/0.0.0-dev/src/S3/S3_File.enso

```diff
@@ -0,0 +1,212 @@
+    read self format=Auto_Detect (on_problems=Problem_Behavior.Report_Warning) =
+        _ = on_problems
+        case format of
+            Auto_Detect -> if self.is_directory then format.read self on_problems else
+                self.if_valid_uri bucket->key->
+                    response = S3.get_object bucket key self.credentials
+                    response.decode Auto_Detect
+            _ -> self.with_input_stream [File_Access.Read] format.read_stream
+
+    ## ALIAS load bytes, open bytes
+       Reads all bytes in this file into a byte vector.
+    read_bytes : Vector ! File_Error
+    read_bytes self =
+        self.read Bytes
+
+    ## ALIAS load text, open text
+       Reads the whole file into a `Text`, with specified encoding.
+
+       Arguments:
+       - encoding: The text encoding to decode the file with. Defaults to UTF-8.
+       - on_problems: Specifies the behavior when a problem occurs during the
+         function.
+         By default, a warning is issued, but the operation proceeds.
+         If set to `Report_Error`, the operation fails with a dataflow error.
+         If set to `Ignore`, the operation proceeds without errors or warnings.
+    @encoding Encoding.default_widget
+    read_text : Encoding -> Problem_Behavior -> Text ! File_Error
+    read_text self (encoding=Encoding.utf_8) (on_problems=Problem_Behavior.Report_Warning) =
+        self.read (Plain_Text encoding) on_problems
+
+    ## GROUP Operators
+       Join two path segments together.
+
+       Arguments:
+       - subpath: The path to join to the path of `self`.
+    / : Text -> S3_File
+    / self subpath = if self.is_directory.not then Error.throw (S3_Error.Error "Only folders can have children." self.uri) else
+        self.if_valid_uri bucket->prefix->
+            if subpath.startsWith "/" then S3_File.Value ("s3://"+bucket+subpath) self.credentials else
+                parts = subpath.split "/"
+                loop current remaining = if remaining.length == 0 then current else
+                    new_current = case remaining.first of
+                        ".." ->
+                            last_index = current.lastIndexOf "/"
+                            if last_index == Nothing then (S3_Error.Error "Cannot move above root folder.") else current.take last_index
+                        "." -> current
+                        x -> new_current + "/" + x
+                    @Tail_Call loop new_current (remaining.drop 1)
+                path = loop prefix parts
+                S3_File.Value ("s3://"+bucket+path) self.credentials
+
+    ## GROUP Calculations
+       Join two or more path segments together, normalizing the `..` and `.` subpaths.
+
+       Arguments:
+       - subpaths: The path segment or segments to join to the path of `self`.
+    join : (Text | Vector) -> S3_File
+    join self subpaths = case subpaths of
+        _ : Vector -> (subpaths.fold self c->p-> c / p) . normalize
```

Comment:
Where is method `normalize` defined for `S3_File`? I cannot find it. It seems that a part of the logic is already included in `/` but I don't see a separate method.

Reply:
It isn't a separate method here as I chose not to allow it.

Reply:
I just don't understand how this works 😅 
I don't see the `normalize` method declared anywhere, so _what_ is called here?

----

But yes, I think exposing it could be useful. And will make the API more consistent with regular `File`, right? That's also a plus IMO.

Reply:
shouldn't have been a normalise here - was a left over.

Reply:
Ah! That's what I was confused about. Makes sense now.